### PR TITLE
Coefficient improvements

### PR DESCRIPTION
**Pull request by mkstoyanov**
Cleanup the coefficient implementation and enable OpenMP.

### CODE REVIEW

**Comment by mkstoyanov** [comment]
@stefan-schnake All tests are passing and I think I fixed the problems with the single-cell, but you'll have to double-check.

This may need a second pass, since your comments are no-longer in the right spot and some of the notation has changed. Nevertheless, I think the current setup is much easier to read and follow.

File: src/coefficients.cpp [hunk 1]
```diff
@@ -235,370 +239,275 @@ fk::matrix<P> generate_coefficients(
...
-      }
-      else if (pterm.coeff_type == coefficient_type::grad ||
-               pterm.coeff_type == coefficient_type::div)
-      {
-        output = legendre_prime_t * tmp * (-1);
+        fk::matrix<P, mem_type::view> blk(coefficients, current,
+                                          current + porder, current,
+                                          current + porder);
+        if constexpr (coeff_type == coefficient_type::mass)
+          fm::gemm(legendre_poly, tmp, blk, true, false, P{1}, P{1});
```
Comment (stefan-schnake) [inline src/coefficients.cpp:323]:
```suggestion
          // volume integral where phi is trial(tmp) and psi is test(legendre_poly)
          //  \int phi(x)\psi(x) dx 
          fm::gemm(legendre_poly, tmp, blk, true, false, P{1}, P{1});
```

File: src/coefficients.cpp [hunk 2]
```diff
@@ -235,370 +239,275 @@ fk::matrix<P> generate_coefficients(
...
+        else // div or grad falls here
+          fm::gemm(legendre_prime, tmp, blk, true, false, P{-1}, P{1});
```
Comment (stefan-schnake) [inline src/coefficients.cpp:326]:
```suggestion
          // -\int \phi(x)\psi'(x) dx
          fm::gemm(legendre_prime, tmp, blk, true, false, P{-1}, P{1});
```

File: src/coefficients.cpp [hunk 3]
```diff
@@ -235,370 +239,275 @@ fk::matrix<P> generate_coefficients(
...
+        // -------------------------------------------------------------------------
+        // More detailed explanation
+        // Each trace_value_ evaluates <FLUX_f,[[v]]>
+        // where v is a DG functions with support on I_i. The
+        // difference between the trace_values_ varies with the edge the flux
+        // is evaluated on and the support of the DG function f.
+        // The legendre_poly_X is the trace of f and legende_poly_X_t is for v
+        // We will use f=p_X for the polynomials where X=L (left boundary of cell)
+        // or X=R (right boundary of cell).  Similar for v but depends on the
+        // support
```
Comment (stefan-schnake) [inline src/coefficients.cpp]:
```suggestion
        // support.  Note matrix multiply ordering goes by
        // v_mat^T * f_mat for <f,v>
```

File: src/coefficients.cpp [hunk 4]
```diff
@@ -235,370 +239,275 @@ fk::matrix<P> generate_coefficients(
...
+
+        // trace_value_1 is the interaction on x_{i-1/2} --
+        // the edge between cell I_{i-1} and I_i or the left boundary of I_i.
+        // f is a DG function with support on I_{i-1}
+        // In this case:  {{f}} = p_R/2, [[f]] = p_R, [[v]] = -p_L
+        coeff_axpy(current, current - porder - 1, -fluxL2 - fluxL2abs, matrix_LtR);
```
Comment (stefan-schnake) [inline src/coefficients.cpp:393]: I believe `(fluxL2 + fluxL2abs)*(-1.0)` is more readable w.r.t. the comments.  Is writing  it this way too slow?

File: src/coefficients.cpp [hunk 5]
```diff
@@ -235,370 +239,275 @@ fk::matrix<P> generate_coefficients(
...
+
+        // trace_value_2 is the interaction on x_{i-1/2} --
+        // the edge between cell I_{i-1} and I_i or the left boundary of I_i.
+        // f is a DG function with support on I_{i}
+        // In this case:  {{f}} = p_L/2, [[f]] = -p_L, [[v]] = -p_L
+        coeff_axpy(current, current, -fluxL2 + fluxL2abs, matrix_LtL);
```
Comment (stefan-schnake) [inline src/coefficients.cpp:399]: Same as above but `(fluxL2 - fluxL2abs)*(-1.0)'

File: src/coefficients.cpp [hunk 6]
```diff
@@ -235,370 +239,275 @@ fk::matrix<P> generate_coefficients(
...
-                (+1);
-          }
-          else
-          {
-            trace_value_3 =
-                (legendre_poly_R_t * legendre_poly_R) * (+1 * flux_right);
-          }
-          trace_value_4 =
-              (legendre_poly_R_t * (legendre_poly_R - legendre_poly_R)) * (+1);
+        case boundary_condition::dirichlet:
```
Comment (stefan-schnake) [inline src/coefficients.cpp:468]:
```suggestion
        case boundary_condition::dirichlet:
        // If penalty then we add <|g|/2[f],[v]>
        // Else we're wanting no flux as this is handed by the
        // boundary conditions.
```

File: src/coefficients.cpp [hunk 7]
```diff
@@ -235,370 +239,275 @@ fk::matrix<P> generate_coefficients(
...
+          if constexpr (coeff_type == coefficient_type::penalty)
+            coeff_axpy(0, 0, fluxL2abs, matrix_LtL);
+          break;
+
+        case boundary_condition::neumann:
```
Comment (stefan-schnake) [inline src/coefficients.cpp:476]:
```suggestion
        case boundary_condition::neumann:
        // If penalty then we add nothing
        // Else we want to standard (outflow) flux
        // <gf,v> = <g{f}/2,v>
```

File: src/coefficients.cpp [hunk 8]
```diff
@@ -235,370 +239,275 @@ fk::matrix<P> generate_coefficients(
...
+          if constexpr (coeff_type != coefficient_type::penalty)
+            coeff_axpy(0, 0, -2.0 * fluxL2, matrix_LtL);
+          break;
+
+        default: // case boundary_condition::periodic
+          coeff_axpy(0, last, -fluxL2 - fluxL2abs, matrix_LtR);
```
Comment (stefan-schnake) [inline src/coefficients.cpp:485]: If you change the computation on the interior fluxes above, change these to match.

File: src/coefficients.cpp [hunk 9]
```diff
@@ -235,370 +239,275 @@ fk::matrix<P> generate_coefficients(
...
+        // For div and grad, the boundary is not part of the bilinear operator,
+        // but instead tranferred to the source.  Similar to an inflow condition.
+        // For penalty, the operator <|gfunc|/2*f,v> is applied for the case where
+        // f and v share the same volume support
+
+        // If statement checking coeff_type is because gfunc can evaluate to nan
+        // in 1/0 case.  Ex: gfunc = x, domain = [0,4] (possible in spherical
+        // coordinates)
+
+        // If neumann
```
Comment (stefan-schnake) [inline src/coefficients.cpp]: Delete this Neumann block chunk.  Keep `//Neumann bondary conditions` onwards.

File: src/coefficients.cpp [hunk 10]
```diff
@@ -235,370 +239,275 @@ fk::matrix<P> generate_coefficients(
...
 
-      if (left == boundary_condition::periodic ||
-          right == boundary_condition::periodic)
-      {
-        if (i == 0)
-        {
-          row1 = current;
-          col1 = last;
+          // left boundary of the right-most cell is in the interior
+          coeff_axpy(last, last - porder - 1, -fluxL2 - fluxL2abs, matrix_LtR);
```
Comment (stefan-schnake) [inline src/coefficients.cpp:512]: Modify flux writing if needed.

File: src/coefficients.cpp (hunk 4 above)
Comment (mkstoyanov) [reply src/coefficients.cpp:393]: Resolved per the face-to-face conversation.

**Review by stefan-schnake — APPROVED** [review]

**Pull request merged**